modules/process/manager_exec_test.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package process
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func retry(t *testing.T, fun func() error, tries int) {
+	var err interface{}
+	for i := 0; i < tries; i++ {
+		err = fun()
+		if err == nil {
+			return
+		}
+		<-time.After(1 * time.Second)
+	}
+	assert.Fail(t, fmt.Sprintf("Retry: failed \n%v", err))
+}
+
+func TestManagerKillGrandChildren(t *testing.T) {
+	tmp := t.TempDir()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	pm := &Manager{
+		processMap: make(map[IDType]*process),
+		next:       1,
+	}
+
+	go func() {
+		// blocks forever because of the firewall at 4.4.4.4
+		_, _, _ = pm.ExecDir(ctx, -1, tmp, "GIT description", "git", "clone", "https://4.4.4.4", "something")
+	}()
+
+	// the git clone process forks a grand child git-remote-https, wait for it
+	pattern := "git-remote-https origin https://4.4.4.4"
+	ps := func() string {
+		cmd := exec.Command("ps", "-x", "-o", "pid,ppid,pgid,args")
```

Comment:
UNIX-only command, tests can also be run on windows machine.

Reply:
Yes. The implementation will need to be different and I'll address that if @zeripath thinks this is a problem that deserves fixing since he recently added a feature to set Gitea children process to be process group leaders.

---
[source](https://lab.forgefriends.org/forgefriends/forgefriends/-/merge_requests/54#note_1488)